### PR TITLE
SC-402 Initial Poetry & Versioning, Publishing

### PR DESCRIPTION
**Pull request by epcAnker**
Validation poetry versioning works:

------

Validation Poetry publishing works (only on new tag)
https://github.com/epcpower/sunspec-demo/actions/runs/1730970110
https://pypi.org/project/sunspecdemo/#history note the 0.1.3 version (this will need to be deleted...)

------

To use:

first install poetry  
for powershell: `(Invoke-WebRequest -Uri https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py -UseBasicParsing).Content | python -`

linux: `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -`

`poetry install` should install dependencies

`poetry run ....` (or activate the .venv with 'poetry env' or run directly from .venv/...)  

---

(will put a poetry guide on confluence)

If you want the .venv to install in your directory:

`poetry config virtualenvs.in-project = true`  
`poetry config virtualenvs.path = "."`

`poetry config --list` shows all config.  mine looks like:

cache-dir = "C:\\\\Users\\\\Alex\\\\AppData\\\\Local\\\\pypoetry\\\\Cache"  
experimental.new-installer = true  
installer.parallel = true  
virtualenvs.create = true  
virtualenvs.in-project = true  
virtualenvs.path = "."

### CODE REVIEW

**Comment by epcAnker** [comment]
individual story is SC-402, will put a comment in jira about this

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -21,9 +21,9 @@ jobs:
       fail-fast: false
       matrix:
         python:
-          - name: CPython 3.9
-            version: 3.9
-            action: 3.9
+          - name: CPython 3.7
+            version: 3.7
+            action: 3.7
```
Comment (gordon-epc) [inline .github/workflows/ci.yml]: Why downgrade from 3.9 to 3.7?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -40,15 +40,19 @@ jobs:
           python-version: ${{ matrix.python.action }}
           architecture: ${{ matrix.arch.action }}
 
-      - name: boots
+      - name: Install poetry ${{ matrix.poetry-version }}
```
Comment (gordon-epc) [inline .github/workflows/ci.yml]: I don't think the `matrix.poetry-version` is working? I don't see it in the log.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -40,15 +40,19 @@ jobs:
...
+          python -m pip install --upgrade pip
+          python -m pip install poetry
 
-      - name: Package executable
+      - name: poetry install
         run: |
-          # help versioneer and pyinstaller get along
-          venv/Scripts/pip install .
-          venv/Scripts/pyinstaller pyinstaller.spec
+          poetry install
```
Comment (gordon-epc) [inline .github/workflows/ci.yml]: Maybe move this under Install poetry? Maybe the wording is just weird. I know it's installing the poetry environment and not installing poetry.

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -61,4 +61,4 @@ Step-by-step release
 
 - Tag the merged commit in the format ``vMAJOR.MINOR.PATCH`` and push the tag to the primary repository.
 
-  - This will result in another build, which will generate the official release artifacts.
+  - This will result in another build, which will generate the official release artifacts.
```
Comment (gordon-epc) [inline CONTRIBUTING.rst]: Is there a difference on this row? Spacing or something?

File: README.rst [hunk 5]
```diff
@@ -23,15 +23,16 @@ Additionally this program acts as a basic example of using the `pysunspec`_ Pyth
...
+pip install poetry-dynamic-versioning
 
 Windows
 =======
 
 ::
 
-    py boots.py ensure
+    poetry install
     venv\Scripts\epcsunspecdemo get-models
```
Comment (gordon-epc) [inline README.rst]: Is the virtual environment directory named `venv`? My limited knowledge suggests it is `.venv` or possibly configurable?

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+[tool.poetry]
+name = "epcsunspecdemo"
+version = "0.1.0.post46.dev0+2a57176"
+description = ""
+authors = ["Alex Anker <alex.anker@epcpower.com>"]
```
Comment (gordon-epc) [inline pyproject.toml:5]: Probably more appropriate to put a generic email address, if we include an `authors` entry at all. Perhaps `info at epcpower`?

File: .gitignore [hunk 7]
```diff
@@ -237,3 +237,6 @@ tags
 [._]*.un~
 
 # End of https://www.gitignore.io/api/eclipse,vim,python,git
+
+# vscode
+.vscode/
```
Comment (gordon-epc) [inline .gitignore]: This should be done with `gitignoreio` instead. Update `gitignore.io` with entry "VisualStudioCode".

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+[tool.poetry-dynamic-versioning]
+enable = true
+
+
+[tool.poetry.dependencies]
+python = "3.7.9"
+attrs = "18.2.0"
+click = "7"
+gitignoreio = "0.1.5"
```
Comment (gordon-epc) [inline pyproject.toml]: Wouldn't `gitignoreio` entry be under `tool.poetry.dev-dependencies`?

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+toolz = "0.9.0"
+tqdm = "4.32.1"
+pysunspec = "2.1.0"
+setuptools = "53.0.0"
+toml = "^0.10.2"
+pyinstaller = "4.2"
```
Comment (gordon-epc) [inline pyproject.toml]: Wouldn't `pyinstaller` be under `tool.poetry.dev-dependencies`? Maybe I am misunderstanding the meaning of "dev-dependencies"?

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (epcAnker) [reply .github/workflows/ci.yml]: i get this error with 3.9

poetry.utils.env.NoCompatiblePythonVersionFound: Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command. 30

ill try to fix it.

Comment (epcAnker) [reply .github/workflows/ci.yml]: copy paste from chat:

so i had downgraded python 3.9 to 3.7.
I am working on getting it back and running for 3.9 but im hitting a block.
so i changed it back to 3.9 in the ci, and in poetry i changed python to ^3.9.
CI runs through and makes the artifact, but the testing of the artifact fails.  I eventually pulled the artifact to test it myself and got this error.
did some research to see that time.clock is deprecated in python 3.8.  Im not sure how this was working before.  I read that the fix is to change time.clock -> time.time.
The issue is that the time.clock is in the .venv, (.venv\Lib\site-packages\sunspec\core\modbus\client.py) so im not exactly sure what to do about that.

File: pyproject.toml (hunk 6 above)
Comment (epcAnker) [reply pyproject.toml:5]: doesnt work without an author

File: CONTRIBUTING.rst (hunk 4 above)
Comment (epcAnker) [reply CONTRIBUTING.rst]: ill revert

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+enable = true
+
+
+[tool.poetry.dependencies]
+python = "^3.7"
+attrs = "18.2.0"
+click = "7"
+toolz = "0.9.0"
+tqdm = "4.32.1"
+pysunspec = "2.1.0"
```
Comment (gordon-epc) [inline pyproject.toml:25]:
```suggestion
pysunspec = "2.1.1"
```

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 6 above)
Comment (gordon-epc) [reply pyproject.toml:5]:
```suggestion
authors = ["EPC Power Corporation <info@epcpower.com>"]
```

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,30 @@
+[tool.poetry]
+name = "epcsunspecdemo"
+version = "0.1.0.post46.dev0+2a57176"
+description = ""
+authors = ["Alex Anker <alex.anker@epcpower.com>"]
+
```
Comment (gordon-epc) [inline pyproject.toml:8]:
```suggestion
license = "MIT"

```

File: pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+[tool.poetry.dev-dependencies]
+gitignoreio = "0.1.5"
+pyinstaller = "4.2"
+
+
+[tool.poetry.scripts]
+epcsunspecdemo = 'epcsunspecdemo.cli:cli'
+
+[build-system]
+requires = ["setuptools", "poetry-core>=1.0.0"]
```
Comment (gordon-epc) [inline pyproject.toml]: In the poetry dynamic versioning docs, I see the following line:

```suggestion
requires = ["setuptools", "poetry-core>=1.0.0", "poetry-dynamic-versioning"]
```

Not sure if this is necessary? Does the dynamic versioning work?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (gordon-epc) [reply .github/workflows/ci.yml]: Appears that updating pysunspec to v2.1.1 has alleviated the issue. Still need to test.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -40,15 +40,16 @@ jobs:
...
+      - name: Install Poetry and Run
         run: |
-          python boots.py ensure --group build
+          python3 -m ensurepip
+          python3 -m pip install --upgrade pip
+          python3 -m pip install poetry
+          poetry install
 
-      - name: Package executable
+      - name: run pyinstaller
```
Comment (gordon-epc) [inline .github/workflows/ci.yml]: Is changing from "Package executable" to "run pyinstaller" necessary? I suppose you could say "Package executable with pyinstaller"

File: .github/workflows/ci.yml (hunk 2 above)
Comment (gordon-epc) [reply .github/workflows/ci.yml]: Looks like you removed the `matrix.poetry-version` from the name.

File: README.rst (hunk 5 above)
Comment (gordon-epc) [reply README.rst]: Looks like it was replaced with `poetry run`.

File: pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+[tool.poetry]
+name = "epcsunspecdemo"
+version = "0.1.0.post46.dev0+2a57176"
```
Comment (gordon-epc) [inline pyproject.toml]: The `poetry-dynamic-versioning` docs suggests:

> Poetry's typical version setting is still required in `[tool.poetry]`, but you are encouraged to use `version = "0.0.0"` as a standard placeholder.

```suggestion
version = "0.0.0"
```

File: .github/workflows/ci.yml (hunk 13 above)
Comment (epcAnker) [reply .github/workflows/ci.yml]: will revert, not sure when i did this.

File: pyproject.toml (hunk 6 above)
Comment (epcAnker) [reply pyproject.toml:5]: spoke about this, will leave in my name and what happens

Comment (gordon-epc) [reply pyproject.toml:5]: Good luck!

File: pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,38 @@
+[tool.poetry]
+name = "epcsunspecdemo"
+version = "0.0.0"
+description = ""
```
Comment (gordon-epc) [inline pyproject.toml]:
```suggestion
description = "EPC SunSpec demonstration tool"
```
This is the description from the now removed `setup.py`.

File: pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+click = "7"
+toolz = "0.9.0"
+tqdm = "4.32.1"
+pysunspec = "2.1.1"
+setuptools = "53.0.0"
+toml = "^0.10.2"
+importlib = "^1.0.4"
+importlib-metadata = "^4.10.1"
+typing-extensions = "^4.0.1"
+[tool.poetry.dev-dependencies]
```
Comment (gordon-epc) [inline pyproject.toml]: Nitpick add empty line.
```suggestion
typing-extensions = "^4.0.1"

[tool.poetry.dev-dependencies]
```

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+gitignoreio = "0.1.5"
+pyinstaller = "4.2"
+poetry-dynamic-versioning = "^0.13.1"
+
+
+[tool.poetry.scripts]
```
Comment (gordon-epc) [inline pyproject.toml]: Nitpick remove extra line
```suggestion
poetry-dynamic-versioning = "^0.13.1"

[tool.poetry.scripts]
```

File: src/sunspecdemo/__init__.py [hunk 18]
```diff
@@ -0,0 +1,2 @@
+#0.0.0 dev build placeholder
```
Comment (gordon-epc) [inline src/sunspecdemo/__init__.py]:
```suggestion
# For development, the __version__ is set to the build placeholder 0.0.0
# For release, the __version__ is the actual released version
```

The comment is sort of confusing when viewed from the perspective of the release package. I suggest this more detailed comment instead.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: .gitattributes [hunk 19]
```diff
@@ -1 +1 @@
-src/epcsunspecdemo/_version.py export-subst
+src/sunspecdemo/_version.py export-subst
```
Comment (gordon-epc) [inline .gitattributes]: The file `src/sunspecdemo/_version.py` no longer exists. Delete `.gitattributes`?

File: MANIFEST.in [hunk 20]
```diff
@@ -1,4 +1,2 @@
 graft src
 LICENSE
-include versioneer.py
-include src/epcsunspecdemo/_version.py
```
Comment (gordon-epc) [inline MANIFEST.in]: According to the [poetry docs](https://github.com/python-poetry/poetry#introduction), the `MANIFEST.in` file is no longer required.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 21]
```diff
@@ -0,0 +1,43 @@
+[tool.poetry]
+name = "sunspecdemo"
+version = "0.1.95.post1.dev0+a38984c"
```
Comment (gordon-epc) [inline pyproject.toml]: Shouldn't this be:
```suggestion
version = "0.0.0"
```

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 21 above)
Comment (epcAnker) [reply pyproject.toml]: yes

**Review by gordon-epc — APPROVED** [review]

File: pyproject.toml [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+[tool.poetry-dynamic-versioning]
+enable = false
+vcs = "git"
+style = "pep440"
+
+[tool.poetry-dynamic-versioning.substitution]
+files = ["src/sunspecdemo/__init__.py"]
+
+[tool.poetry.dependencies]
+python = ">=3.7.9, <4.0"
```
Comment (gordon-epc) [inline pyproject.toml]:
```suggestion
python = ">=3.7.9, <4.0"
```
```suggestion
python = "^3.7.9"
```
These are equivalent statements.

**Review by gordon-epc — CHANGES_REQUESTED** [review]

**Review by gordon-epc — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 23]
```diff
@@ -39,16 +42,35 @@ jobs:
...
+
+          poetry install
 
       - name: Package executable
         run: |
-          # help versioneer and pyinstaller get along
-          venv/Scripts/pip install .
-          venv/Scripts/pyinstaller pyinstaller.spec
+          # set the versioning
+          poetry run poetry-dynamic-versioning.exe
```
Comment (gordon-epc) [inline .github/workflows/ci.yml]:
```suggestion
          poetry run poetry-dynamic-versioning
```
```suggestion
          poetry run poetry-dynamic-versioning.exe
```

Comment (gordon-epc) [inline .github/workflows/ci.yml:71]:
```suggestion
          poetry run poetry-dynamic-versioning
```
```suggestion
          poetry run poetry-dynamic-versioning.exe
```

Comment (gordon-epc) [inline .github/workflows/ci.yml:71]:
```suggestion
          poetry run poetry-dynamic-versioning
```
```suggestion
          poetry run poetry-dynamic-versioning.exe
```

**Review by gordon-epc — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 23 above)
Comment (gordon-epc) [reply .github/workflows/ci.yml]: Not sure why it put the suggested change twice?

**Review by gordon-epc — APPROVED** [review]

**Pull request merged**